### PR TITLE
OJ-1655 (OJ-1406) add audit events to Favourite Lambda

### PR DESCRIPTION
**Pull request by scronk1**
### What changed

Added the functionality to send two audit events from the favourite lambda:
REQUEST_SENT &
RESPONSE_RECEIVED

### Why did it change

In order to check the events with TXMA they need to be emitting and these two are a boilerplate approach when we're making a third party API requests

### Issue tracking

- [OJ-1406](https://govukverify.atlassian.net/browse/OJ-1406) - general txma ticket
- [OJ-1655](https://govukverify.atlassian.net/browse/OJ-1655) - subtask to add events to favourite lambda

### Environment variables or secrets

- [x] No environment variables or secrets were added or changed

[OJ-1406]: https://govukverify.atlassian.net/browse/OJ-1406?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[OJ-1655]: https://govukverify.atlassian.net/browse/OJ-1655?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: lambdas/favourite/src/favourite-handler.ts [hunk 1]
```diff
@@ -42,7 +48,19 @@ export class FavouriteLambda implements LambdaInterface {
...
+
+    const auditEventContext = this.createAuditEventContext(sessionItem);
+
+    await this.auditService.sendAuditEvent(
+      AuditEventType.REQUEST_SENT,
+      auditEventContext
+    );
     const response = await this.callExternalToy(toyBody.toy);
+    await this.auditService.sendAuditEvent(
+      AuditEventType.THIRD_PARTY_REQUEST_ENDED,
```
Comment (surajkumar) [inline lambdas/favourite/src/favourite-handler.ts]: It needs to be `RESPONSE_RECEIVED` as we no longer use `THIRD_PARTY_REQUEST_ENDED`

**Review by surajkumar — CHANGES_REQUESTED** [review]

File: lambdas/favourite/src/favourite-handler.ts (hunk 1 above)
Comment (scronk1) [reply lambdas/favourite/src/favourite-handler.ts]: Ah ok! I was going by what was in the TS AuditEventType but I will update

File: lambdas/types/audit-event.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+export interface AuditEventSession {
+  sessionId: string;
+  subject: string;
+  persistentSessionId: string;
+  clientSessionId: string;
+}
+
+export interface AuditEventContext {
```
Comment (surajkumar) [inline lambdas/types/audit-event.ts:19]: Should there be headers too? Like the one in common? 
https://github.com/alphagov/di-ipv-cri-lib/blob/main/src/main/java/uk/gov/di/ipv/cri/common/library/domain/AuditEventContext.java

Comment (surajkumar) [reply lambdas/types/audit-event.ts:19]: In KBV: `auditService.sendAuditEvent(AuditEventType.RESPONSE_RECEIVED,new AuditEventContext(requestHeaders, sessionItem),`

Comment (scronk1) [reply lambdas/types/audit-event.ts:19]: Resolved in slack - as I'm matching the following structure https://govukverify.atlassian.net/wiki/spaces/OJ/pages/3607330894/KBV+Audit+Event+Payload it's not hugely important how that's done in the lambda

File: lambdas/common/config/config-service.ts [hunk 3]
```diff
@@ -51,6 +52,25 @@ export class ConfigService {
     }
   }
 
+  public getAuditConfig(): CriAuditConfig {
```
Comment (jkunle) [inline lambdas/common/config/config-service.ts]: I know this was the original design, but I wonder if we need `CriAuditConfig` the only thing that come from config is the issuer. Perhaps the bulk of this should move into the Audit code and then we just pass in the issuer.

Comment (scronk1) [reply lambdas/common/config/config-service.ts]: Yeah makes sense, will move it into the audit service

**Review by jkunle — APPROVED** [review]

File: lambdas/favourite/src/favourite-handler.ts [hunk 4]
```diff
@@ -25,13 +25,19 @@ import { CommonConfigKey } from "../../types/config-keys";
...
 import createAuthorizationCodeMiddleware from "../../middlewares/session/create-authorization-code-middleware";
+import { AuditService } from "../../common/services/audit-service";
+import { SQSClient } from "@aws-sdk/client-sqs";
+import { AuditEventContext, AuditEventType } from "../../types/audit-event";
 
 const dynamoDbClient = createClient(AwsClientType.DYNAMO) as DynamoDBDocument;
 const ssmClient = createClient(AwsClientType.SSM) as SSMClient;
+const sqsClient = createClient(AwsClientType.SQS) as SQSClient;
 const FAVOURITE_METRIC = "favourite_selected";
 const TOY_API_URL = process.env.TOY_API_URL;
```
Comment (surajkumar) [inline lambdas/favourite/src/favourite-handler.ts:37]: We should probably have some sort of error handling for when the env is not present

File: lambdas/favourite/src/favourite-handler.ts [hunk 5]
```diff
@@ -42,7 +48,33 @@ export class FavouriteLambda implements LambdaInterface {
...
+    );
+
+    await this.auditService.sendAuditEvent(
+      AuditEventType.REQUEST_SENT,
+      auditEventContext
+    );
     const response = await this.callExternalToy(toyBody.toy);
+
+    const toyResponse =
+      (response.status !== 200 ? "ToyNotFoundError: " : "Toy found: ") +
```
Comment (surajkumar) [inline lambdas/favourite/src/favourite-handler.ts]: What about other status codes? Maybe we should flip the condition

**Review by surajkumar — CHANGES_REQUESTED** [review]

File: lambdas/favourite/src/favourite-handler.ts (hunk 4 above)
Comment (surajkumar) [reply lambdas/favourite/src/favourite-handler.ts:37]: TOY_API_URL *

File: lambdas/favourite/src/favourite-handler.ts (hunk 5 above)
Comment (scronk1) [reply lambdas/favourite/src/favourite-handler.ts]: Seems there's only 200 and 404 defined - https://github.com/alphagov/di-ipv-third-party-stubs/blob/e2104ea4cc01b7edb277fa73af0530bc024805d0/stubs/toy/config/openapi-spec.yaml#L19, but I guess you can always get a 500 or 403 so I'll add specificity for those 👍

Comment (surajkumar) [reply lambdas/favourite/src/favourite-handler.ts]: Consider `if(200) { toy found } else 404 { toynotfound } else { error }`

File: lambdas/favourite/src/favourite-handler.ts [hunk 6]
```diff
@@ -60,23 +61,39 @@ export class FavouriteLambda implements LambdaInterface {
...
+        toyResponse: "",
       },
     };
-    await this.auditService.sendAuditEvent(
-      AuditEventType.RESPONSE_RECEIVED,
-      receivedContext
-    );
 
-    if (response.status !== 200) {
+    if (statusCode == 200) {
```
Comment (surajkumar) [inline lambdas/favourite/src/favourite-handler.ts:66]:
```
const statusCode = response.status;

function updateContextAndSendAuditEvent(msg) {
  const receivedContext = {
    ...auditEventContext,
    extensions: {
      toy: toyBody.toy,
      toyResponse: msg,
    },
  };
  this.auditService.sendAuditEvent(AuditEventType.RESPONSE_RECEIVED, receivedContext);
}

if (statusCode === 200) {
  const msg = `${toyBody.toy} found: received 200 response`;
  updateContextAndSendAuditEvent(msg);
  logger.info(msg);
} else if (statusCode === 404) {
  const msg = "ToyNotFoundError: received 404 response";
  updateContextAndSendAuditEvent(msg);
  throw new ToyNotFoundError(toyBody.toy);
} else {
  const msg = `Error: received ${statusCode} response - ${response.statusText}`;
  updateContextAndSendAuditEvent(msg);
  throw new GenericServerError(msg);
}
```
This should remove the duplication

**Review by surajkumar — APPROVED** [review]
LGTM

**Pull request merged**